src/main/kotlin/com/bloock/sdk/record/entity/document/Document.kt

```diff
@@ -0,0 +1,124 @@
+package com.bloock.sdk.record.entity.document
+
+import com.bloock.sdk.proof.entity.Proof
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.runBlocking
+
+abstract class Document<T>(src: T, args: DocumentLoadArgs) {
+    var ready: Deferred<Unit>
+    protected var data: T? = null
+    protected var payload: T? = null
+    protected var signatures: Signatures? = Signatures(mutableListOf())
+    protected var proof: Proof? = null
+
+    init {
+        runBlocking {
+            ready = GlobalScope.async {
+                setup(src).await()
+                proof = fetchProof()
+                signatures = fetchSignatures().await()
+                data = fetchData().await()
+                payload = fetchPayload().await()
+
+            }
+        }
+    }
+
+    abstract suspend fun fetchData(): Deferred<T?>
+
+    suspend fun fetchSignatures(): Deferred<Signatures> {
+        return GlobalScope.async {
+            return@async fetchMetadata("signatures").await()?.let {
+                if (it != null) {
+                    it as MutableList<Any>
+
+                    return@let Signatures(it)
+                }
+                Signatures(mutableListOf())
+
+            } ?: Signatures(mutableListOf())
+        }
+    }
+
+    suspend fun fetchProof(): Proof? {
+        var proof = fetchMetadata("proof").await()
+        return proof?.let {
+            return proof as Proof
+
+        }
+
+    }
+
+    protected suspend abstract fun setup(src: T): Deferred<Unit>
+    protected suspend abstract fun fetchMetadata(key: String): Deferred<Any?>
+    protected suspend fun fetchPayload(): Deferred<T> {
+        var metadata = mutableMapOf<String, Any>()
+        this.signatures?.value?.let {
+            if (it.isNotEmpty()) {
+
+                metadata.put("signatures", this.signatures!!)
+            }
+        }
+
+        return this.buildFile(metadata)
+    }
+
+
+    suspend fun build(): Deferred<T> {
+        var metadata = SignatureWithProof(null, mutableListOf())
+        this.proof?.let {
+            metadata.proof = this.proof
+        }
+        this.signatures?.let {
+            it.value?.let {
+
+                metadata.signatures = it
+            }
+        }
+
+        return mapOf(Pair("signatures", metadata.signatures))?.let { this.buildFile(it) }
```

Comment:
The proof isn't added to the output. Also, it isn't loaded in the metadata initializer (L70).